Loki appender didn't retry on connect timeout

Hi,

we have a connect timeout from time to time at the startup of the application.

shouldn't he have retried in this case ?
as it retry in case of ConnectException is seem logical for me to retried here as well, we know for sure the logs wasn't send.

```
21:01:21,387 |-ERROR in com.github.loki4j.client.pipeline.AsyncBufferPipeline@1624588e - Error while sending Batch #35dbe3f1b75483 (480 bytes) to Loki (https://xxx/loki/api/v1/push) java.net.http.HttpConnectTimeoutException: HTTP connect timed out
	at java.net.http.HttpConnectTimeoutException: HTTP connect timed out
	at 	at java.net.http/jdk.internal.net.http.HttpClientImpl.send(HttpClientImpl.java:567)
	at 	at java.net.http/jdk.internal.net.http.HttpClientFacade.send(HttpClientFacade.java:123)
	at 	at com.github.loki4j.client.http.JavaHttpClient.send(JavaHttpClient.java:68)
	at 	at com.github.loki4j.client.pipeline.AsyncBufferPipeline.sendBatch(AsyncBufferPipeline.java:322)
	at 	at com.github.loki4j.client.pipeline.AsyncBufferPipeline.sendStep(AsyncBufferPipeline.java:293)
	at 	at com.github.loki4j.client.pipeline.AsyncBufferPipeline.runSendLoop(AsyncBufferPipeline.java:224)
	at 	at com.github.loki4j.client.pipeline.AsyncBufferPipeline.lambda$start$3(AsyncBufferPipeline.java:131)
	at 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at 	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.net.http.HttpConnectTimeoutException: HTTP connect timed out
	at 	at java.net.http/jdk.internal.net.http.ResponseTimerEvent.handle(ResponseTimerEvent.java:68)
	at 	at java.net.http/jdk.internal.net.http.HttpClientImpl.purgeTimeoutsAndReturnNextDeadline(HttpClientImpl.java:1270)
	at 	at java.net.http/jdk.internal.net.http.HttpClientImpl$SelectorManager.run(HttpClientImpl.java:899)
Caused by: java.net.ConnectException: HTTP connect timed out
	at 	at java.net.http/jdk.internal.net.http.ResponseTimerEvent.handle(ResponseTimerEvent.java:69)
	at 	... 2 common frames omitted
```

loki-logback-appender 1.5.1

java -version:
```
openjdk version "17.0.11" 2024-04-16
OpenJDK Runtime Environment (build 17.0.11+9-alpine-r0)
OpenJDK 64-Bit Server VM (build 17.0.11+9-alpine-r0, mixed mode, sharing)
```

Regards.

--- a/loki-client/src/main/java/com/github/loki4j/client/pipeline/AsyncBufferPipeline.java
+++ b/loki-client/src/main/java/com/github/loki4j/client/pipeline/AsyncBufferPipeline.java
@@ -1,6 +1,7 @@
 package com.github.loki4j.client.pipeline;
 
 import java.net.ConnectException;
+import java.net.http.HttpTimeoutException;
 import java.util.Comparator;
 import java.util.Optional;
 import java.util.concurrent.ConcurrentLinkedQueue;
@@ -370,7 +371,9 @@ public final class AsyncBufferPipeline {
     }
 
     private boolean checkIfEligibleForRetry(Exception e, LokiResponse r) {
-        return e instanceof ConnectException || (r != null && checkIfStatusEligibleForRetry(r.status));
+        return e instanceof ConnectException
+                || e instanceof HttpTimeoutException
+                || (r != null && checkIfStatusEligibleForRetry(r.status));
     }
 
     private boolean checkIfStatusEligibleForRetry(int status) {

--- a/loki-logback-appender/src/test/java/com/github/loki4j/logback/Loki4jAppenderTest.java
+++ b/loki-logback-appender/src/test/java/com/github/loki4j/logback/Loki4jAppenderTest.java
@@ -301,6 +301,42 @@ public class Loki4jAppenderTest {
     }
 
     @Test
+    public void testHttpTimeoutRetry() throws InterruptedException, BrokenBarrierException, TimeoutException, ExecutionException {
+        StringPayload expectedPayload = StringPayload.builder()
+            .stream("[level, INFO, app, my-app]",
+                "ts=100 l=INFO c=test.TestApp t=thread-1 | Test message 1 ")
+            .build();
+
+        var failingHttpClient = new FailingHttpClient();
+        var sender = new WrappingHttpSender<>(failingHttpClient);
+        var encoder = defaultToStringEncoder();
+
+        var appender = appender(1, 4000L, encoder, sender);
+        appender.setMinRetryBackoffMs(50);
+        appender.setMaxRetryJitterMs(10);
+        appender.start();
+
+        sender.client.setFailure(FailureType.HTTP_CONNECT_TIMEOUT_EXCEPTION);
+        var sendCapture = sender.client.captureSendInvocation();
+
+        appender.append(events[0]);
+
+        var send1 =  sendCapture.waitForNextSend(100);
+        assertEquals("send", 1, send1.sendNo);
+        assertEquals("send", expectedPayload, StringPayload.parse(send1.data, encoder.charset));
+
+        var send2 = send1.waitForNextSend(150);
+        assertEquals("retry1", 2, send2.sendNo);
+        assertEquals("retry1", expectedPayload, StringPayload.parse(send2.data, encoder.charset));
+
+        var send3 = send2.waitForNextSend(200);
+        assertEquals("retry2", 3, send3.sendNo);
+        assertEquals("retry2", expectedPayload, StringPayload.parse(send3.data, encoder.charset));
+
+        appender.stop();
+    }
+
+    @Test
     public void testRateLimitedRetry() throws InterruptedException, BrokenBarrierException, TimeoutException, ExecutionException {
         StringPayload expectedPayload = StringPayload.builder()
             .stream("[level, INFO, app, my-app]",

--- a/testkit/src/main/java/com/github/loki4j/testkit/dummy/FailingHttpClient.java
+++ b/testkit/src/main/java/com/github/loki4j/testkit/dummy/FailingHttpClient.java
@@ -1,6 +1,7 @@
 package com.github.loki4j.testkit.dummy;
 
 import java.net.ConnectException;
+import java.net.http.HttpConnectTimeoutException;
 import java.nio.ByteBuffer;
 
 import com.github.loki4j.client.http.HttpStatus;
@@ -14,7 +15,9 @@ public class FailingHttpClient extends DummyHttpClient {
     public LokiResponse send(ByteBuffer batch) throws Exception {
         var response = super.send(batch);
         if (failureType == FailureType.CONNECTION_EXCEPTION)
-            throw new ConnectException("Text exception");
+            throw new ConnectException("Text ConnectException");
+        else if (failureType == FailureType.HTTP_CONNECT_TIMEOUT_EXCEPTION)
+            throw new HttpConnectTimeoutException("Test HttpConnectTimeoutException");
         else if (failureType == FailureType.RATE_LIMITED)
             return new LokiResponse(HttpStatus.TOO_MANY_REQUESTS, "Rate Limited Request");
         return response;
@@ -27,6 +30,7 @@ public class FailingHttpClient extends DummyHttpClient {
     public enum FailureType {
         NONE,
         CONNECTION_EXCEPTION,
+        HTTP_CONNECT_TIMEOUT_EXCEPTION,
         RATE_LIMITED
     }
 

